### PR TITLE
[Fix #15969] Throw error when Popover content is unexpected type

### LINKED ISSUE #15969: Popover is not shown if content is 0 (numeric)
We use popover to show some counts/stats, so it is not surprising we just put a number in there.
However 0 doesn't show up -- probably because it's falsy.

Example:
http://jsfiddle.net/fL5s3hmp/5/ (press Test-0 to see the problem).

### PR DESCRIPTION
**Pull request by kkirsche**
[Fix #15969] Throw error when Popover content is unexpected type

### CODE REVIEW

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 75faeb0628f0414ddc532d1d1929c371bda908ca
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52745293

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: b1952e989528e02d1447effc2fe728e21c8e600c
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52753016

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 0c01635fce8a4f17d88161e76cbacee2210d3622
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52755488

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/popover.js [hunk 1]
```diff
@@ -65,10 +65,16 @@
     var $e = this.$element
     var o  = this.options
 
-    return $e.attr('data-content')
+    var content = $e.attr('data-content')
       || (typeof o.content == 'function' ?
             o.content.call($e[0]) :
             o.content)
+
+    if (['[object Function]', '[object Object]', '[object String]'].indexOf(Object.prototype.toString.call(content)) === -1) {
```
Comment (cvrebert) [inline js/popover.js]: How about we just use http://api.jquery.com/jquery.isfunction/ for the function check?

**Comment by cvrebert** [comment]
* `title` needs the same typechecking
* The Tooltip plugin needs the same typechecking

File: js/popover.js (hunk 1 above)
Comment (kkirsche) [inline js/popover.js]: Any reason for this over the standard javascript library?

Comment (cvrebert) [inline js/popover.js]: I mean, this approach is kinda hacky, less easy to read, and also seems to incorrectly allow plain objects (e.g. `{}`)

Comment (kkirsche) [inline js/popover.js]: We have to since tests send object object

**Comment by kkirsche** [comment]
Is title also function / string type?

**Comment by cvrebert** [comment]
Yes, just as the docs say.

File: js/popover.js (hunk 1 above)
Comment (cvrebert) [inline js/popover.js]: Hmm? What do you mean?

Comment (kkirsche) [inline js/popover.js]: At the point that I'm type checking I get the jQuery object as [object Object]. That's the reason for me using deep checking here instead of typeOf

Comment (cvrebert) [inline js/popover.js]: Should be able to use `instanceof $` for that: http://stackoverflow.com/a/1853246/3342739

Comment (hnrch02) [inline js/popover.js]: Am I missing something or why isn't something like `if (/string|function/.test(typeof content))` used here? Also, according to the docs a jQuery object isn't even allowed for the `content` option.

Comment (cvrebert) [inline js/popover.js]: 5bf13914e49942409c0b21aeaf2ed756c41bd498 implies that we do support it, although there are no tests and it's missing from the docs as you mentioned.

Comment (cvrebert) [inline js/popover.js]: #14552 is supposed to add proper support for DOM and jQuery objects here

Comment (kkirsche) [inline js/popover.js]: @hnrch02 when setting the variable content in this, I get the value of the function as `content` not the function itself. Thus, in the tests I end up with a jQuery object from items like:

```js
content: function () {
          return $div
        }
```

Comment (kkirsche) [inline js/popover.js]: Ah, I shouldn't grab content there, I need to test before on `o.content`. Fixing this :+1:

**Comment by kkirsche** [comment]
This should be much cleaner / less hacky now @cvrebert | @hnrch02 

Let me know what you think. I was testing content after it had been called, I needed to do it before then. This is also now implemented on Tooltip / titles (by implementing it on Tooltip, it covers the titles for both tooltip and popover).

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: be69b68dc0b43ef0e5969eed638b59904c0d1abe
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52900706

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/popover.js [hunk 2]
```diff
@@ -65,6 +65,10 @@
     var $e = this.$element
     var o  = this.options
 
+    if (!/string|function/.test(typeof o.content)) {
+      throw new Error('Popover received a(n) ' + typeof o.content + '. Expected a string or a function.')
```
Comment (hnrch02) [inline js/popover.js]: Maybe just rephrase this to `Popover received content of type ${type}. Was expecting a string or a function.` or something like that to remove the need for "a(n)".

Comment (kkirsche) [inline js/popover.js]: Fixed @hnrch02. Thanks for the suggestion :smiley:. I agree that wording works better.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 1279f112684ae7f01db6eda69a8976abea2dfe35
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52952774

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
If given a function, when we call it, we should check that it returned a string.

**Comment by kkirsche** [comment]
@cvrebert do we want to check it's a string and change [our tests](https://github.com/twbs/bootstrap/blob/master/js/tests/unit/popover.js#L89) or do we want to check that it returned a string or jQuery / html object and leave the existing tests as they are?

**Comment by kkirsche** [comment]
OK, so in 2 tests I am having issues with it not being a String.

`should detach popover content rather than removing it so that event handlers are left intact` and `should not duplicate HTML object`. I get `[object Object]` with constructor `function (a, b) {return new m.fn.init(a,b);}`. Even though I get back `$('<div class="content-with-handler"><a class="btn btn-warning">Button with event handler</a></div>').appendTo('#qunit-fixture')`, `o.content.call($e[0]) instanceof $` doesn't work. Still trying to identify how to handle this.

**Comment by kkirsche** [comment]
@cvrebert Type checking of the function return value has been added.

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 682f2a9ff680494ef35696f9669d9ecaa8858be7
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/52969616

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by hnrch02** [comment]
@twbs-savage retry

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 682f2a9ff680494ef35696f9669d9ecaa8858be7
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53003064

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/popover.js [hunk 3]
```diff
@@ -64,9 +64,22 @@
   Popover.prototype.getContent = function () {
     var $e = this.$element
     var o  = this.options
+    var contentType = typeof o.content
+
+    if (!/string|function/.test(contentType)) {
+      throw new Error('Popover received type ${' + contentType + '}. Was expecting type string or a function.')
```
Comment (hnrch02) [inline js/popover.js]: `${}` are part of the ES6 interpolation syntax. I was just trying to make clear where the variable should be included :smile:

**Comment by hnrch02** [comment]
Maybe we could abstract the type-checking into a shared function between tooltip and popover?

File: js/popover.js [hunk 4]
```diff
@@ -64,9 +64,22 @@
...
+    }
+
+    if (contentType == 'function') {
+      var contentReturnType = typeof o.content.call($e[0])
+      if (!/string/.test(contentReturnType) && !(o.content.call($e[0]) instanceof $)) {
```
Comment (hnrch02) [inline js/popover.js]: The RegExp doesn't make sense if you only have one thing to compare against.

Comment (hnrch02) [inline js/popover.js]: Also, not good that the content function is called twice.

File: js/tests/unit/popover.js [hunk 5]
```diff
@@ -232,4 +232,62 @@ $(function () {
...
+  })
+
+  QUnit.test('should throw an error when popover content function returns a non-string object', function (assert) {
+    assert.expect(1)
+    var $popover = $('<a href="#">@fat</a>')
+      .appendTo('#qunit-fixture')
+      .bootstrapPopover({
+        title: 'Test',
+        content: function () {
+          return new Date();
```
Comment (hnrch02) [inline js/tests/unit/popover.js]: Semicolon.

File: js/popover.js (hunk 4 above)
Comment (hnrch02) [inline js/popover.js]: Also<sup>2</sup>, this doesn't handle DOM nodes only jQuery object. Plus it's less heavy if you check for a jQuery object by using `returnValueOfContentFunction.jquery`, see http://api.jquery.com/jquery-2/.

File: js/tests/unit/popover.js (hunk 5 above)
Comment (kkirsche) [inline js/tests/unit/popover.js]: Oops! Sorry about that. It's a habit that sneaks in sometimes

File: js/popover.js (hunk 4 above)
Comment (kkirsche) [inline js/popover.js]: You are 100% correct. Fixing this now.

File: js/popover.js (hunk 3 above)
Comment (kkirsche) [inline js/popover.js]: Oops… Sorry! I admit I need to read more about ES6. Thanks for clearing that up. :flushed:

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 3ceaa8b9db53220c7fa301f82b20cb6b745c8aa1
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53053713

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by kkirsche** [comment]
@hnrch02 Fixed all of the comments you had and refactored into methods. :+1:

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 299db1fb4979ebfe6f0f7fe06e2b6a65219eb521
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53055135

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by kkirsche** [comment]
Wanted to follow up and see if this was missing anything :smiley:

File: js/tooltip.js [hunk 6]
```diff
@@ -385,13 +385,36 @@
...
   }
 
+  Tooltip.prototype.didntReceiveTypeStringOrFunction = function (received) {
+    return !/string|function/.test(received)
+  }
+
+  Tooltip.prototype.didntReturnStringOrJQuery = function (returned, returnedType) {
+    return (returnedType !== 'string' && !(returned.jquery))
+  }
+
```
Comment (cvrebert) [inline js/tooltip.js]: Personally, I would move the `throw`s into these functions and rename them along the lines of `checkValidContentType`

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: dbb21fbac68ca54ec96ffaf3391de0e65f5d8866
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53539429

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by kkirsche** [comment]
Refactored further @cvrebert per your suggestions

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 9c7f15a1cbc36c41dc28f52649369001a88d8743
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/53539870

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by kkirsche** [comment]
Wanted to follow up on this and check if it was missing anything.

**Comment by kkirsche** [comment]
Wanted to follow up about this again.

**Comment by nextgenthemes** [comment]
@kkirsche u think spamming will get u anywhere?

**Comment by cvrebert** [comment]
@nextgenthemes Nothing wrong with an occasional ping on a PR. Nobody wants their contribution/time to go to waste.

**Comment by kkirsche** [comment]
Wanted to follow up on this and check if it was missing anything or if it was waiting for fat / someone else to review it. Thanks for your time.

**Comment by cvrebert** [comment]
@twbs-savage Retry

**Comment by cvrebert** [comment]
@twbs-savage Retry

File: js/tooltip.js [hunk 7]
```diff
@@ -385,13 +385,54 @@
...
 
+  Tooltip.prototype.didModuleReceiveTypeStringOrFunction = function (module, received) {
+    if (!/string|function/.test(received)) {
+      throw new Error(module + ' received type ' + received + '. Was expecting type string or a function.')
+    }
+
+    return true
+  }
+
+  Tooltip.prototype.didFunctionReturnStringOrJQuery = function (module, returned, returnedType) {
```
Comment (hnrch02) [inline js/tooltip.js]: I'm still not happy with these method names. They aren't used anywhere else so why not just inline them into `checkValidTypes`?

Comment (hnrch02) [inline js/tooltip.js]: This doesn't handle normal DOM nodes being returned which would also be added by #14552.

File: js/tooltip.js [hunk 8]
```diff
@@ -385,13 +385,54 @@
...
+    var returnedContent
+    this.didModuleReceiveTypeStringOrFunction(module, contentType)
+
+    if (contentType == 'function') {
+      if (module === 'Popover') {
+        returnedContent = o.content.call(this.$element[0])
+      }
+      if (module === 'Tooltip') {
+        returnedContent = o.title.call(this.$element[0])
+      }
```
Comment (hnrch02) [inline js/tooltip.js]: This can be simplified into `returnedContent = o[module == 'Popover' ? 'content' : 'title'].call(this.$element[0])`.

Comment (hnrch02) [inline js/tooltip.js]: Also, as far as I understand it, #14552 allows both Popover titles and the content to be a string / function / jQuery object / DOM node not only the content.

Comment (hnrch02) [inline js/tooltip.js]: And what about directly passing in a DOM node or a jQuery object instead of a function returning either of those?

File: js/tooltip.js [hunk 9]
```diff
@@ -385,13 +385,54 @@
...
+  Tooltip.prototype.didFunctionReturnStringOrJQuery = function (module, returned, returnedType) {
+    if (returnedType !== 'string' && !(returned.jquery)) {
+      throw new Error(module + ' received type function which returned type ' + returnedType + '. ' +
+        'Was expecting type string as the return type.')
+    }
+
+    return true
+  }
+
+  Tooltip.prototype.checkValidType = function (module, contentType, o) {
```
Comment (hnrch02) [inline js/tooltip.js]: `module` is unnecessary, use `this.type`.

Comment (kkirsche) [inline js/tooltip.js]: Fixed.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: dd9a622de58915e50c8c3f0752ee70fda655cf4b
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58993735

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/tooltip.js (hunk 8 above)
Comment (kkirsche) [inline js/tooltip.js]: Fixed I believe.

File: js/tooltip.js (hunk 7 above)
Comment (kkirsche) [inline js/tooltip.js]: Fixed I believe.

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: c48794b64ea5f0e3a7ced28a8527e07f0b2cc5f7
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58994379

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by kkirsche** [comment]
Yeah yeah, I know I know @twbs-savage 

Pull request updated to address, I believe, all of the mentioned issues @hnrch02 

I apologize for the delay. Started a new job so been doing transition related stuff the last month. Working on fixing tests now

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 393669d4a315b6b71efbd74aeee9e2896cef5f88
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58995022

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 8bd34c559a4b462c29edb37f9cb266d03ff11f56
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58996547

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: 1c730c1281ac4d6ea06baecf39165ee1c6bfa770
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/58996582

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by kkirsche** [comment]
Nope...this is all kinds of messed up after all the commits since when this was originally done :disappointed:.

If you could let me know the @twbs team's interest level is so I can know whether this is of interest / importance to the team and I should invest the time to rewrite it and fix all the things that changed or if this is just another feature which is nice but nobody on the team honestly cares about.

**Comment by fat** [comment]
hey @kkirsche thanks a ton for all of this work – sorry we didn't get it in.

Our currently plan is to write some more generalized type checker for v4.

So i'm going to close this for now, but will reference your work (and maybe pull in your unit tests).

Thanks again!

**Comment by kkirsche** [comment]
It's ok. I understand. I hope some of it can be helpful though for you guys!

**Comment by mdo** [comment]
<3 @kkirsche
